src/debugger.rs

```diff
@@ -0,0 +1,218 @@
+        writeln!(self.output.borrow_mut(), "{formatted_string}").map_err(|e| e.to_string())?;
+        flush()?;
+        Ok(())
+    }
+
+    fn build_cli(&self) -> Command {
+        let mut cli = Command::new("repl")
+            .multicall(true)
+            .arg_required_else_help(true)
+            .subcommand_required(true)
+            .subcommand_value_name("DEBUGGER")
+            .subcommand_help_heading("IXA DEBUGGER")
+            .help_template("{all-args}");
+
+        for (name, handler) in &self.commands {
+            cli = cli.subcommand(Command::new(*name).about(handler.about()).help_template(
+                "{about-with-newline}\n{usage-heading}\n    {usage}\n\n{all-args}{after-help}",
+            ));
+        }
+
+        cli
+    }
+}
+
+/// Returns the current population of the simulation
+struct PopulationCommand;
+impl DebuggerCommand for PopulationCommand {
+    fn about(&self) -> &'static str {
+        "Get the total number of people"
+    }
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        _matches: &ArgMatches,
+    ) -> Result<bool, String> {
+        cli.writeln(&format!("{}", context.get_current_population()))?;
+        Ok(false)
+    }
+}
+
+/// Exits the debugger and continues the simulation
+struct ContinueCommand;
+impl DebuggerCommand for ContinueCommand {
+    fn about(&self) -> &'static str {
+        "Continue the simulation and exit the debugger"
+    }
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        _matches: &ArgMatches,
+    ) -> Result<bool, String> {
+        cli.writeln(&format!(
+            "Continuing the simulation from t = {}",
+            context.get_current_time()
+        ))?;
+        Ok(true)
+    }
+}
+
+fn readline() -> Result<String, String> {
+    write!(std::io::stdout(), "$ ").map_err(|e| e.to_string())?;
+    std::io::stdout().flush().map_err(|e| e.to_string())?;
+    let mut buffer = String::new();
+    std::io::stdin()
+        .read_line(&mut buffer)
+        .map_err(|e| e.to_string())?;
+    Ok(buffer)
+}
+
+fn setup_repl(line: &str, context: &mut Context) -> Result<bool, String> {
+    let args = shlex::split(line).ok_or("error: Invalid quoting")?;
+
+    let mut repl = DebuggerRepl::new();
+    repl.register_command("population", Box::new(PopulationCommand));
+    repl.register_command("continue", Box::new(ContinueCommand));
+    let matches = repl
+        .build_cli()
+        .try_get_matches_from(args)
+        .map_err(|e| e.to_string())?;
+
+    if let Some((command, sub_matches)) = matches.subcommand() {
+        // If the provided command is known, run its handler
+        if let Some(handler) = repl.get_command(command) {
+            return handler.handle(context, &repl, sub_matches);
+        }
+        // Unexpected command: print an error
+        return Err(format!("Unknown command: {command}"));
+    }
+
+    unreachable!("subcommand required");
```

Comment:
This can never happen, right? Because you register commands above.

Reply:
correct this is just to handle the unhandled else case which should in practice never be hit